src/components/content/heroes/aurora/weather-easter-egg/precipitation-effect.tsx

```diff
@@ -0,0 +1,185 @@
+import React, { useEffect, useState } from 'react';
+import styled, { keyframes } from 'styled-components';
+
+interface RaindropProps {
+  color: string;
+  speed: number;
+  blur: number;
+  left: number;
+  top: number;
+  animation_delay: number;
+}
+
+interface SnowflakeProps {
+  speed: number;
+  size: number;
+  animation: any;
+  blur: number;
+  transparency: number;
+  left: number;
+  top: number;
+  animation_delay: number;
+}
+
+interface SnowflakeContainerProps {
+  speed: number;
+  top: number;
+  left: number;
+  animation_delay: number;
+}
+
+interface PrecipitationData {
+  id: number;
+  color: string;
+  speed: number;
+  delay: number;
+  left: number;
+}
+
+const fallAnimation = keyframes`
+  0% {
+    transform: translateY(-75px);
+    opacity: 0;
+  }
+  100% {
+    transform: translateY(150vh);
+    opacity: 1;
+  }
+`;
+
+const generateSwayAnimation = (initialX: number, range: number) => keyframes`
+  0% {
+    transform: translateX(0);
+  }
+  50% {
+    transform: translateX(${initialX + Math.random() * range - range / 2}px);
+  }
+  100% {
+    transform: translateX(${
+      initialX + (Math.random() * range) / 2 - range / 4
+    }px);
+  }
+`;
+
+const Raindrop = styled.div<RaindropProps>`
+  position: absolute;
+  width: 4px;
+  height: 50px;
+  border-radius: 150%;
+  background-color: ${(props) => props.color};
+  animation: ${fallAnimation} ${(props) => props.speed}s linear infinite;
+  filter: blur(${(props) => props.blur}px);
+  left: ${(props) => props.left}%;
+  top: ${(props) => props.top}px;
+  animation-delay: ${(props) => props.animation_delay}s;
+`;
+
+const Snowflake = styled.div<SnowflakeProps>`
+  animation-delay: ${(props) => props.animation_delay}s;
+  top: ${(props) => props.top}px;
+  left: ${(props) => props.left}%;
+  position: absolute;
+  width: ${(props) => props.size}px;
+  height: ${(props) => props.size}px;
+  border-radius: 150%;
+  animation: ${(props) => props.animation} ${(props) => props.speed / 3}s
+    ease-in-out infinite alternate;
+  background: radial-gradient(
+    50% 50% at 50% 50%,
+    rgba(255, 255, 255, ${(props) => props.transparency}) 0%,
+    rgba(255, 255, 255, ${(props) => props.transparency}) 20%,
+    rgba(255, 255, 255, ${(props) => props.transparency * 0.8}) 85%,
+    rgba(255, 252, 222, ${(props) => props.transparency}) 100%
+  );
+  filter: blur(${(props) => props.blur}px);
+`;
+
+const SnowflakeContainer = styled.div<SnowflakeContainerProps>`
+  animation: ${fallAnimation} ${(props) => props.speed}s linear infinite;
+  position: absolute;
+  top: ${(props) => props.top}px;
+  left: ${(props) => props.left}%;
+  animation-delay: ${(props) => props.animation_delay}s;
+`;
+
+export enum PrecipitationType {
+  Rain = 'rain',
+  Snow = 'snow',
+}
+
+const PrecipitationEffect = ({ dropCount, speed, type, speedDeviation }) => {
```

Comment:
I appreciate your generalization efforts to reuse this for rain & snow. 

Followggn some thoughts about the level of abstraction you chose. Actually nothing to change but maybe something to think about for you 👍

This function looks like you went the generalization path to early (kind of a premature optimization) as this effect looks more like a bundle of two effects that share some parameters and the little code around "generateDrops". 

I see a function with two big conditions and few more. Your code would be much easier to read and more importantly to experiment with and to derive other experiments if you have a dedicated RainEffect and SnowEffect instead of merging both into a single effect. Those effects could even include their own copy of `generateDrop` for the beginning. Doesn't actually matter and simple to share once you reach that point.